File: Lua/jean_parser.lua
local JeanParser = {}

local function is_whitespace(c)
    return c == " " or c == "\t" or c == "\n" or c == "\r"
end

function JeanParser.parse_jean_data(text)
    local tokens = {}
    local i = 1
    local n = #text

    while i <= n do
        local char = text:sub(i, i)

        if is_whitespace(char) then
            i = i + 1

        elseif char == "%" then
            while i <= n and text:sub(i, i) ~= "\n" and text:sub(i, i) ~= "\r" do
                i = i + 1
            end

        elseif char == '"' then
            local str_start = i + 1
            i = i + 1
            while i <= n and text:sub(i, i) ~= '"' do
                i = i + 1
            end

            if i > n then
                io.stderr:write("JEAN Parser: Unterminated string literal\n")
                return tokens
            end

            table.insert(tokens, text:sub(str_start, i - 1))
            i = i + 1

        else
            local word_start = i
            while i <= n and not is_whitespace(text:sub(i, i)) and text:sub(i, i) ~= "%" do
                i = i + 1
            end

            local word = text:sub(word_start, i - 1)

            if #word > 0 then
                if word == "t" then
                    table.insert(tokens, true)
                elseif word == "f" then
                    table.insert(tokens, false)
                else
                    local num = tonumber(word)
                    if num ~= nil then
                        table.insert(tokens, num)
                    else
                        io.stderr:write("JEAN Parser: Unrecognized token '" .. word .. "'\n")
                    end
                end
            end
        end
    end

    return tokens
end

function JeanParser.parse_jean_file(file_path)
    local file, err = io.open(file_path, "r")
    if not file then
        io.stderr:write("JEAN Parser: Could not open file - " .. tostring(err) .. "\n")
        return {}
    end

    local content = file:read("*a")
    file:close()

    return JeanParser.parse_jean_data(content)
end

return JeanParser

File: Python/jean_parser.py
def parse_jean_data(text: str) -> list:
    tokens = []
    i = 0
    n = len(text)

    while i < n:
        char = text[i]

        if char.isspace():
            i += 1
            continue

        if char == '%':
            while i < n and text[i] not in '\r\n':
                i += 1
            continue

        if char == '"':
            start = i + 1
            i += 1
            while i < n and text[i] != '"':
                i += 1
            if i >= n:
                raise ValueError("Unterminated string literal")
            tokens.append(text[start:i])
            i += 1
            continue

        start = i
        while i < n and not text[i].isspace() and text[i] != '%':
            i += 1
        word = text[start:i]

        if word == 't':
            tokens.append(True)
        elif word == 'f':
            tokens.append(False)
        else:
            try:
                tokens.append(int(word))
            except ValueError:
                try:
                    tokens.append(float(word))
                except ValueError:
                    raise ValueError(f"Unrecognized token: '{word}'")

    return tokens


def parse_jean_file(file_path: str) -> list:
    with open(file_path, 'r', encoding='utf-8') as f:
        file_content = f.read()
    return parse_jean_data(file_content)